src/server/json_family.cc

```diff
@@ -89,83 +134,294 @@ OpResult<string> OpGet(const OpArgs& op_args, string_view key, vector<pair<strin
+  if (!result) {
+    return result.status();
+  }
 
-  if (!decoder.is_valid()) {
-    return OpStatus::SYNTAX_ERR;
+  vector<OptSizeT> vec;
+  auto cb = [&vec](const string_view& path, const json& val) {
+    if (val.is_object()) {
+      vec.emplace_back(val.object_value().size());
+    } else {
+      vec.emplace_back(nullopt);
+    }
+  };
+
+  expression.evaluate(*result, cb);
+  return vec;
+}
+
+OpResult<vector<OptSizeT>> OpArrLen(const OpArgs& op_args, string_view key, JsonExpression expression) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
   }
 
   vector<OptSizeT> vec;
   auto cb = [&vec](const string_view& path, const json& val) {
-    if (val.is_string()) {
-      vec.emplace_back(val.as_string_view().size());
+    if (val.is_array()) {
+      vec.emplace_back(val.array_value().size());
     } else {
       vec.emplace_back(nullopt);
     }
   };
 
-  expression.evaluate(decoder.get_result(), cb);
+  expression.evaluate(*result, cb);
+  return vec;
+}
+
+OpResult<vector<OptBool>> OpToggle(const OpArgs& op_args, string_view key, string_view path) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
+
+  vector<OptBool> vec;
+  auto cb = [&vec](const string& path, json& val) {
+    if (val.is_bool()) {
+      bool current_val = val.as_bool() ^ true;
+      val = current_val;
+      vec.emplace_back(current_val);
+    } else {
+      vec.emplace_back(nullopt);
+    }
+  };
+
+  json j = result.value();
+  error_code ec = JsonReplace(j, path, cb);
+  if (ec) {
+    VLOG(1) << "Failed to evaulate expression on json with error: " << ec.message();
+    return OpStatus::SYNTAX_ERR;
+  }
+
+  SetString(op_args, key, j.as_string());
+  return vec;
+}
+
+OpResult<vector<OptLongLong>> OpNumIncrBy(const OpArgs& op_args, string_view key, string_view path, long long num) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
+
+  vector<OptLongLong> vec;
+  auto cb = [&](const string& path, json& val) {
+    if (val.is_number()) {
+      auto current_val = val.as<long long>() + num;
+      val = current_val;
+      vec.emplace_back(current_val);
+    } else {
+      vec.emplace_back(nullopt);
+    }
+  };
+
+  json j = result.value();
+  error_code ec = JsonReplace(j, path, cb);
+  if (ec) {
+    VLOG(1) << "Failed to evaulate expression on json with error: " << ec.message();
+    return OpStatus::SYNTAX_ERR;
+  }
+
+  SetString(op_args, key, j.as_string());
+  return vec;
+}
+
+OpResult<vector<OptLongLong>> OpNumMultBy(const OpArgs& op_args, string_view key, string_view path, long long num) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
+
+  vector<OptLongLong> vec;
+  auto cb = [&](const string& path, json& val) {
+    if (val.is_number()) {
+      auto current_val = val.as<long long>() * num;
```

Comment:
I should return an OVERFLOW error if the result is out of the range of 64-bit IEEE double.

Reply:
In order to provide fully compatible behavior, we need to setup an elasticache instance and learn how these special cases are handled there. That's out of scope of this assignment at this point.

what I usually do is I write `// TODO: .... ` comments when the code needs further enhancement.